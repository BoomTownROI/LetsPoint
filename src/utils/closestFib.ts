/**
 * Finds the closest average fibonacci number
 * 0 and under are removed
 */
export default function closestFib(votes: number[]): number {
  const validVotes = votes.filter(v => v > 0).filter(v => isFinite(v));
  const average = validVotes.reduce((prev, v) => prev + v, 0) / validVotes.length;
  if (isNaN(average) || average === 0) return 0;

  // Find two fib number that bound the value
  let fib0 = 0;
  let fib1 = 1;
  while (!isBound(fib0, average, fib1)) {
    const nextFib1 = fib1 + fib0;
    fib0 = fib1;
    fib1 = nextFib1;

    if (nextFib1 > 1000) {
      throw new Error(`You are spinning on calculating nextFib: votes: ${JSON.stringify(votes)}`);
    }
  }

  // Which Fib number are we closer to
  const isCloserToFib0 = average - fib0 < fib1 - average;
  return isCloserToFib0 ? fib0 : fib1;
}

function isBound(lower: number, val: number, upper: number): boolean {
  return lower <= val && val <= upper;
}
