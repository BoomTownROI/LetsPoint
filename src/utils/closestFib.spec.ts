import closestFib from './closestFib';

describe('@closestFib', () => {
  it('Can calculate the average vote closest to a fib. number', () => {
    // average: 1
    expect(closestFib([1, 1, 1, 1])).toBe(1);

    // average: 3
    expect(closestFib([3, 3, 3])).toBe(3);

    // average: 2.3333333333
    expect(closestFib([3, 3, 1])).toBe(2);

    // average: 4.2
    expect(closestFib([1, 5, 5, 5, 5])).toBe(5);

    // average: 7.2
    expect(closestFib([5, 8, 13, 5, 5])).toBe(8);
  });

  it('Errors out when things get out of control', () => {
    expect(() => {
      closestFib([100000000000000000000]);
    }).toThrowErrorMatchingSnapshot();
  });

  it('Does not count zeros when calculating things', () => {
    expect(closestFib([0, 0, 0, 0, 13])).toBe(13);
  });

  it("Doesn't blow up when the values are crazy", () => {
    // Ugly or unexpected
    expect(closestFib([])).toBe(0);
    expect(closestFib([0, 0, 0])).toBe(0);

    // Unlikely, but not caught by ts
    expect(closestFib([-1, -100])).toBe(0);
    expect(closestFib([Infinity])).toBe(0);
    expect(closestFib([Infinity])).toBe(0);
    expect(closestFib([-Infinity])).toBe(0);
    expect(closestFib([-Infinity, Infinity])).toBe(0);
  });
});
