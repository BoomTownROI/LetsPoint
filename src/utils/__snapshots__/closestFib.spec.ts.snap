// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@closestFib Errors out when things get out of control 1`] = `"You are spinning on calculating nextFib: votes: [100000000000000000000]"`;
